### PR TITLE
*: functional-tester with v3

### PR DESCRIPTION
**Pull request by gyuho**
For https://github.com/coreos/etcd/issues/4380.

### CODE REVIEW

**Comment by xiang90** [comment]
I feel this complicates stuff. Let's just keep them separate binary?

**Comment by gyuho** [comment]
Ok will separate commands as it was.

**Comment by gyuho** [comment]
@xiang90 PTAL.

1. commit with minor clean ups.
2. commit with v3 kv stressers.

Thanks!

File: tools/functional-tester/etcd-tester/stresser.go [hunk 1]
```diff
@@ -42,6 +44,85 @@ type stresser struct {
...
+	for i := 0; i < s.N; i++ {
+		go func(i int) {
+			for {
+				putctx, putcancel := context.WithTimeout(ctx, clientV2.DefaultRequestTimeout)
+				kvc := clients[i%len(clients)]
+				_, err := kvc.Put(putctx, &pb.PutRequest{
+					Key:   []byte(fmt.Sprintf("foo%d", rand.Intn(s.KeySuffixRange))),
+					Value: []byte(randStr(s.KeySize)),
+				})
+				putcancel()
```
Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: @xiang90 This is pretty much the same as in v2 stresser. PTAL.

Thanks,

File: tools/functional-tester/etcd-tester/stresser.go [hunk 2]
```diff
@@ -42,6 +44,85 @@ type stresser struct {
...
+
+func (s *stresser) Stress() error {
+	// TODO: make this configurable to take advantage of HTTP2 streams
+	connsN := 1
+	clientsN := 10
+
+	var conns []*grpc.ClientConn
+	for i := 0; i < connsN; i++ {
+		// TODO: use clientv3 for redialing, etc.
+		conn, err := grpc.Dial(s.Endpoint, grpc.WithInsecure())
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: dial with timeout please.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 3]
```diff
@@ -42,6 +44,85 @@ type stresser struct {
...
+	cancel func()
+}
+
+func (s *stresser) Stress() error {
+	// TODO: make this configurable to take advantage of HTTP2 streams
+	connsN := 1
+	clientsN := 10
+
+	var conns []*grpc.ClientConn
+	for i := 0; i < connsN; i++ {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: let's just get one connection.

Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: let's do one connection + one client.

Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: Ok

File: tools/functional-tester/etcd-tester/stresser.go (hunk 2 above)
Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: Ok will fix.

File: tools/functional-tester/etcd-tester/cluster.go [hunk 4]
```diff
@@ -110,8 +126,20 @@ func (c *cluster) Bootstrap() error {
 		}
 	}
 
-	stressers := make([]Stresser, len(clientURLs))
-	for i, u := range clientURLs {
+	var stressers []Stresser
+	if c.v2Only {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/cluster.go]: if c.v20only is ture, we will go into this if statement and then we will continue the for loop at line 142. so we will start both v2 and v3. this is not what we want.

Comment (gyuho) [inline tools/functional-tester/etcd-tester/cluster.go]: Oh right. Need to fix that. Thanks.

**Comment by gyuho** [comment]
@xiang90 

1. Passing grpc.Timeout in the dial option
2. Add `else` statement for `v2only` flag, so that we either run v2 or v3.
3. 1 conn, 1 client for grpc connection

Please let me know if you have any feedback. Thanks.

**Comment by gyuho** [comment]
And in the next commit, I need to implement setHealthKey using v3. (forgot to include that)

File: tools/functional-tester/etcd-tester/stresser.go [hunk 5]
```diff
@@ -42,6 +44,85 @@ type stresser struct {
...
+
+	mu      sync.Mutex
+	failure int
+	success int
+
+	cancel func()
+}
+
+func (s *stresser) Stress() error {
+	// TODO: make this configurable to take advantage of HTTP2 streams
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: we probably do not want to do this. we can add more stresser as we want. each stresser is just a simple unit.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 6]
```diff
@@ -42,6 +44,85 @@ type stresser struct {
...
+	connsN := 1
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: let's remove this paras. keep each stresser simple.

File: tools/functional-tester/etcd-tester/cluster.go [hunk 7]
```diff
@@ -198,15 +233,36 @@ func (c *cluster) Status() ClusterStatus {
 // setHealthKey sets health key on all given urls.
 func setHealthKey(us []string) error {
 	for _, u := range us {
-		cfg := etcdclient.Config{
+		cfg := clientv2.Config{
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/cluster.go]: add a todo?

File: tools/functional-tester/etcd-tester/stresser.go (hunk 6 above)
Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: Ok will just use 1 conn, 1 client.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 8]
```diff
@@ -42,6 +44,85 @@ type stresser struct {
...
+	clientsN := 1
+
+	var conns []*grpc.ClientConn
+	for i := 0; i < connsN; i++ {
+		// TODO: use clientv3 for redialing, etc.
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: we will not do redialing for the stresser probably. we just let it die.

Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: Ok

**Comment by xiang90** [comment]
LGTM

**Comment by gyuho** [comment]
Thanks, will merge after green lights. I will share the doc about logging proposal shortly.

**Pull request merged**